### ISSUE TITLE
Performance Regression in `require` in v2

### ISSUE THREAD

**Issue by mhassan1** [DataDog/dd-trace-js#1847]
**Describe the bug**
We've noticed that our server startup time has slowed from 20 seconds to 100 seconds since we upgraded to `dd-trace@2`. This is true whether or not `tracer.init()` is called, so we believe it could be related to [this change](https://github.com/DataDog/dd-trace-js/pull/1644/files#diff-4385230ff878fca193d8f340227ec177feffbbc2f270322819fddbe0c21cc7d0R37).

To reproduce:
```js
const start = Date.now()

require('dd-trace')

for (let i = 0; i < 5e3; i++) {
  const pkg = 'bson' // or any large package that's installed
  require(pkg)
  delete require.cache[require.resolve(pkg)]
}

console.log(Date.now() - start)
```

Without `dd-trace`: 11 seconds
With `dd-trace@1`: 11 seconds
With `dd-trace@2`: 53 seconds

**Environment**

* **Operation system:** OS X
* **Node version:** 16.11.1
* **Tracer version:** 2.2.0
* **Agent version:** N/A

**Comment by bengl** [comment]
I'm working on a solution in https://github.com/DataDog/dd-trace-js/tree/bengl/fix-ritm-recursion, but currently it's breaking large swaths of tests. Once that's resolved we'll be able to do a release soon after.

**Comment by rochdev** [comment]
Just a quick update that we're still working on it. I also updated the above benchmark to call `init()` since we now patch modules on import instead of on init, and while it's still slightly slower than before, it should be much better than the 53 seconds you were seeing. Hopefully it should be in the <15s range.

**Comment by rochdev** [comment]
Here are the numbers I'm getting so far with the ongoing work:

* Without `dd-trace`: 4.4 seconds
* With `dd-trace@1`: 5.4 seconds
* With `dd-trace@2`: 46 seconds
* With the fix: 5.6 seconds

We have changed a lot of things with how we instrument things which is why it took a bit of time to get this to the finish line, but it's almost done. It depends on another PR that should be merged soon, then we should be able to ship the fix in the next few days.

**Comment by dusan-dragon** [comment]
@rochdev Please this also affects our applications and also developer experience is really terrible, we updated just couple of days ago and it is bad....
Do you have any estimations when this can be fixed?

**Comment by dusan-dragon** [comment]
Also just to illustrate how bad this is... For our dev setup it went from 22s -> 10m :(

**Comment by rochdev** [comment]
We've just merged #1804 which was the prerequisite for fixing this with the new plugin system in place. I'll take another look tomorrow and report back.

**Comment by dusan-dragon** [comment]
@rochdev any update? Any ETA? Trying to evaluate if we really need to downgrade to I guess v1

**Comment by rochdev** [comment]
@dusan-dragon Goal would be early next week.

**Comment by rochdev** [comment]
The fix was released in 2.4.1.
